mailet/standard/src/test/java/org/apache/james/transport/matchers/HasExceptionTest.java

```diff
@@ -0,0 +1,128 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.matchers;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.mailet.Mail;
+import org.apache.mailet.Matcher;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMatcherConfig;
+import org.apache.mailet.base.test.MailUtil;
+import org.junit.Before;
+import org.junit.Test;
+
+public class HasExceptionTest {
+
+    private FakeMail mockedMail;
+    private Matcher matcher;
+
+    @Before
+    public void setUp() throws Exception {
+        MimeMessage mimeMessage = MailUtil.createMimeMessage();
+        mockedMail = MailUtil.createMockMail2Recipients(mimeMessage);
+        matcher = new HasException();
+    }
+
+    @Test
+    public void matchShouldReturnAddressesWhenSpecifiedExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new javax.mail.internet.AddressException());
+
+        FakeMatcherConfig matcherConfig = FakeMatcherConfig.builder().matcherName("HasException")
+                .condition("javax.mail.internet.AddressException").build();
+
+        matcher.init(matcherConfig);
+
+        assertThat(matcher.match(mockedMail)).containsExactlyElementsOf(mockedMail.getRecipients());
+    }
+
+    @Test
+    public void matchShouldReturnAddressesWhenSubclassOfSpecifiedExceptionHasOccurred() throws MessagingException {
+        mockedMail.setAttribute(Mail.MAILET_ERROR_ATTRIBUTE_NAME, new javax.mail.internet.AddressException());
+
+        FakeMatcherConfig matcherConfig = FakeMatcherConfig.builder().matcherName("HasException")
+                .condition("javax.mail.MessagingException").build();
```

Comment:
You can add a similar test with a short class name (i.e. "MessagingException" instead of "javax.mail.MessagingException").

Reply:
I disagree with that one.

We have an interest keeping the tests as close as possible to real world existing class, so that we **can use tests as actual documentation**. That is the reason why I did no remark to these Fully Qualified class names.

Reply:
Nothing prevent us from a user to define such configuration in _mailetcontainer.xml_
```
<mailet match="HasException=ManageSieveException" class="...>
```
so it may be a good idea to cover this by a test.

Reply:
Sorry, I misunderstood you ;-)

Of course.